crates/http-api-bindings/src/completion/openai.rs

```diff
@@ -40,47 +41,94 @@ impl OpenAICompletionEngine {
     }
 }
 
-#[derive(Serialize)]
+#[derive(Debug, Default, Serialize)]
 struct CompletionRequest {
     model: String,
-    prompt: String,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
+    prompt: Option<String>,
+
+    #[serde(skip_serializing_if = "Option::is_none")]
     suffix: Option<String>,
+
     max_tokens: i32,
     temperature: f32,
     stream: bool,
     presence_penalty: f32,
+
+    #[serde(skip_serializing_if = "Vec::is_empty")]
+    messages: Vec<Message>,
+}
+
+#[derive(Serialize, Debug)]
+struct Message {
+    role: String,
+    content: String,
 }
 
 #[derive(Deserialize)]
 struct CompletionResponseChunk {
     choices: Vec<CompletionResponseChoice>,
 }
 
-#[derive(Deserialize)]
+#[derive(Deserialize, Default)]
 struct CompletionResponseChoice {
-    text: String,
+    text: Option<String>,
+    delta: Option<CompletionResponseDelta>,
     finish_reason: Option<String>,
 }
 
+#[derive(Clone, Deserialize, Default, Debug)]
+struct CompletionResponseDelta {
+    content: Option<String>,
+}
+
 #[async_trait]
 impl CompletionStream for OpenAICompletionEngine {
     async fn generate(&self, prompt: &str, options: CompletionOptions) -> BoxStream<String> {
-        let (prompt, suffix) = if self.support_fim {
-            split_fim_prompt(prompt)
-        } else {
-            (prompt, None)
-        };
+        const LEGACY_MODEL_NAME: &str = "gpt-3.5-turbo-instruct";
 
-        let request = CompletionRequest {
+        let mut request = CompletionRequest {
             model: self.model_name.clone(),
-            prompt: prompt.to_owned(),
-            suffix: suffix.map(|x| x.to_owned()),
             max_tokens: options.max_decoding_tokens,
             temperature: options.sampling_temperature,
             stream: true,
             presence_penalty: options.presence_penalty,
+            ..Default::default()
         };
 
+        if self.model_name == LEGACY_MODEL_NAME {
+            let (prompt, suffix) = if self.support_fim {
+                split_fim_prompt(prompt)
+            } else {
+                (prompt, None)
+            };
+            request.prompt = Some(prompt.into());
+            request.suffix = suffix.map(Into::into);
```

Comment:
For other fields marked as `#[serde(skip_serializing_if = "Option::is_none")]`, will be ignored while serializing to json.